### DISCUSSION TITLE
OAuth2 does create a record but cannot login user

### DISCUSSION THREAD

**Discussion by MuslemRahimi** [pocketbase/pocketbase#2772] [Q&A] [answered]
Hi,

I have the problem that when I use google oauth that the user is getting created correctly but when i try to login it does not work.

ONLY when i change the password in my pocketbase Admin UI to something different the user can be logged in.

Anyone knows what the problem is?

I use form with the oauth2 action:
```
oauth2: async ( { cookies, url, locals }) => {
		const authMethods = await locals.pb?.collection('users').listAuthMethods();

		if (!authMethods) {
            return {
                authProviderRedirect: '',
                authProviderState: ''
            };
        }
        const redirectURL = `${url.origin}/oauth`;
        const googleAuthProvider = authMethods.authProviders[0];
        const authProviderRedirect = `${googleAuthProvider.authUrl}${redirectURL}`;
        const state = googleAuthProvider.state;
        const verifier = googleAuthProvider.codeVerifier;

		oauthState.update( value => state);
        oauthVerifier.update( value => verifier);


		throw redirect(302,authProviderRedirect)

	}
```

My route /oauth with +server.ts:

```
import { redirect } from '@sveltejs/kit';
import { oauthState, oauthVerifier } from '$lib/store';
import { generateUsername } from '$lib/utils';

export const GET = async ({locals,  url}) => {
    //console.log(url.searchParams);
    const redirectURL = `${url.origin}/oauth`;
    //const expectedState = cookies.get('state');
    //const expectedVerifier = cookies.get('verifier');
    let newUser;
    let oauthUsername;

    let expectedState;
    let expectedVerifier;

    oauthState.subscribe(value => {
        expectedState = value;
    })
    oauthVerifier.subscribe(value => {
        expectedVerifier = value;
    })

    //console.log('expected',expectedState)

    const state = await url.searchParams.get('state');
    const code = await url.searchParams.get('code');

    //console.log('returned state',state)
    //console.log('returned code',code)

    //as a side effect this will generate a new code verifier, hence why we need to pass the verifier back in through the cookie
    const authMethods = await locals.pb?.collection('users').listAuthMethods();

    
    if (!authMethods?.authProviders) {
        console.log('No Auth Providers');
        throw redirect(303, '/register');
    }
    const provider = authMethods.authProviders[0];
    if (!provider) {
        console.log('Provider Not Found');
        throw redirect(303, '/register');
    }

    if (expectedState !== state) {
        console.log('Returned State Does not Match Expected', expectedState, state);
        throw redirect(303, '/register');
    }

    try {
        //console.log(provider)
        /*
        newUser = await locals.pb?.collection('users')
            .authWithOAuth2Code(provider.name, code, expectedVerifier, redirectURL,{username:'',name:'My Awesome User'});
        */
            newUser = await locals.pb?.collection('users')
            .authWithOAuth2Code(provider.name, code, expectedVerifier, redirectURL);

        oauthUsername = generateUsername(newUser['meta']['name'].split(' ').join('')).toLowerCase();
        
        await locals.pb?.collection('users').update(
            newUser['record']['id'], {
                'username' : oauthUsername
            });

    } catch (err) {
        console.log('Error logging in with OAuth2 user', err);
    }

    //Login user automatically

    console.log('==============================')
    //const avatarUrl = newUser['meta']['avatarUrl'];
    //console.log(avatarUrl);

    throw redirect(303, "/");

};
```

**Comment by ganigeorgiev** [comment, marked as answer]
What do you mean by _"getting created correctly but when i try to login it does not work"_?

Does the `catch` returns any error?

Are you sure that the 303 redirects back with the auth cookie (see https://github.com/pocketbase/pocketbase/discussions/903#discussioncomment-4369424)?

**Reply by MuslemRahimi** [reply]
No catch error was returned.

Ah i see in my hooks.server.ts i set:

`
response.headers.set('set-cookie', event.locals.pb.authStore.exportToCookie({ sameSite: 'lax', secure: false }));
`
Then the user was logged in automatically without problems. 
However when i log out and try to l log in the user again in the old fashion way of email adress and password i get the error "Failed to authenticate".

Solution:
Change the password manually in the admin ui and suddenly the login works again.

I dont know what to do with this bug

**Reply by ganigeorgiev** [reply]
> However when i log out and try to l log in the user again in the old fashion way of email adress and password i get the error "Failed to authenticate".

I'm confused. Are you providing a password during the OAuth2 auth record creation?

**Reply by MuslemRahimi** [reply]
No, i thought that the password of google from that particular user will be used in the pocketbase database no?

**Reply by MuslemRahimi** [reply]
sorry i think i misunderstood how oauth2 works. So in general only specific data will be shared from the provider but not the password of the user obviously. My bad ..

**Reply by ganigeorgiev** [reply]
> So in general only specific data will be shared from the provider but not the password of the user obviously.

Yes. We don't have access to the OAuth2 account password and that would be a security issue in general. The OAuth2 sign-in/sign-up (_if you are not already authenticated_) happens in the provider page, and we get only a success/failure redirect response with a temp access code that we exchange for token. This token is used for getting the OAuth2 profile data (eg. email, unique id, name, etc.). All available OAuth2 data to us are returned in the success auth response under the `meta` prop (eg. in your case `newUser.meta`).

**Comment by MuhammadKhizar7** [comment]
I also have an issue. Although the user record is created, certain fields such as name are not updated with the information available in the social authentication response from Google. This results in incomplete user information being stored in the system.

**Reply by ganigeorgiev** [reply]
The only record fields we autoset with the OAuth2 record creation are the `email` (if available), and the `username` (if available and comply with the PocketBase username format, otherwise - autogenerate).

Any other record fields need to be updated in a separate `update()` call because they are dynamic and user defined. The OAuth2 user data can be accessed in the OAuth2 success response under the `meta` key. _Eventually sometime in the future we may provide an option to map specific OAuth2 provider fields to the user defined record fields_.
